include/secp256k1.h

```diff
@@ -701,6 +713,172 @@ SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ec_pubkey_combine(
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(4);
+
+/** Tweak the private key of an x-only pubkey by adding a tweak to it. The public
+ *  key of the resulting private key will be the same as the output of
+ *  secp256k1_xonly_pubkey_tweak_add called with the same tweak and corresponding
+ *  input public key.
+ *
+ *  If the public key corresponds to a point with square Y, tweak32 is added to
+ *  the seckey (modulo the group order). Otherwise, tweak32 is added to the
+ *  negation of the seckey (modulo the group order).
+ *
+ *  Returns: 1 if the tweak was successfully added to seckey
+ *           0 if the tweak was out of range or the resulting private key would be
+ *             invalid (only when the tweak is the complement of the private key) or
+ *             seckey is 0.
+ *
+ *  Args:      ctx: pointer to a context object, initialized for signing (cannot be NULL)
+ *  In/Out: seckey: pointer to a 32-byte private key
+ *  In:    tweak32: pointer to a 32-byte tweak
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_xonly_privkey_tweak_add(
+    const secp256k1_context* ctx,
+    unsigned char *seckey,
+    const unsigned char *tweak32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Tweak an x-only public key by adding tweak times the generator to it.
+ *
+ *  Because the resulting point may have a non-square Y coordinate, it may not
+ *  be representable by an x-only pubkey. Instead, `output_pubkey` will be set
+ *  to the negation of that point. Therefore this function outputs `is_negated`
+ *  which is required for `xonly_pubkey_tweak_test`.
+ *
+ *  Returns: 1 if tweak times the generator was successfully added to pubkey
+ *           0 if the tweak was out of range or the resulting public key would be
+ *             invalid (only when the tweak is the complement of the corresponding
+ *             private key).
+ *
+ *  Args:           ctx: pointer to a context object initialized for validation
+ *                       (cannot be NULL)
+ *  Out:  output_pubkey: pointer to a public key object (cannot be NULL)
+ *           is_negated: pointer to an integer that will be set to 1 if
+ *                       `output_pubkey` is the negation of the point that
+ *                       resulted from adding the tweak. (cannot be NULL)
+ *  In: internal_pubkey: pointer to an x-only public key object to apply the
+ *                       tweak to (cannot be NULL)
+ *              tweak32: pointer to a 32-byte tweak (cannot be NULL)
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_xonly_pubkey_tweak_add(
+    const secp256k1_context* ctx,
+    secp256k1_xonly_pubkey *output_pubkey,
+    int *is_negated,
+    const secp256k1_xonly_pubkey *internal_pubkey,
+    const unsigned char *tweak32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4) SECP256K1_ARG_NONNULL(5);
```

Comment:
How do people feel about promising that it's fine for `output_pubkey` and `internal_pubkey` to point to the same object?
Right now it will work. but a lot of our APIs start by zeroing out the output object. and then this isn't fine anymore.

On one hand it's nice to not return uninitialized objects. on the other a lot of times you want to tweak the key in place.

Reply:
> Right now it will work. but a lot of our APIs start by zeroing out the output object. and then this isn't fine anymore.

I don't see why. `secp256k1_ec_pubkey_tweak_add` zeroes the key and is in place. Actually we should also zeroize the output key in this function.

Also, that `xonly_pubkey_tweak_add` does not work in place (and therefore differs from `ec_pubkey_tweak_add`) is just for historical reasons (initially output_pubkey was a different type than internal_pubkey, namely `secp256k1_pubkey`). How do people feel about changing the function to tweak the pubkey in-place? I think I prefer that because otherwise it's inconsistent with pubkey_tweak_add.

Reply:
Okay, `tweak_add` changes the pubkey in-place now. And added a test that the pubkey is zeroed on failure.

Reply:
The reason for zeroizing the outputs is mostly so that if the caler doesn't know errors need to be handled, that the output is invalid so that hopefully later error handling will do something useful.  This is especially a concern because errors should never be seen without crafted inputs, and a plausible caller simply won't know they're possible at all. This isn't incompatible with in-place operation. (though not-in-place operation is arguably riskier to not zero because it might expose the content of secrets if the caller ignores the error).

Zeroizing also makes programming errors more likely to get detected faster, e.g. if the function is caled with an invaid output pointer but also an invalid input,  if it zeroizes it's more likely to cause a valgrind error and/or program crash rather than silently being broken.